CCDB: ensure the right number of bytes is deleted for libuv handlers

knopers8 opened the pull request:
libuv handlers use c-style polymorphism, so deleting uv_handle_t* pointing to a derived class will in fact delete wrong number of bytes. As ugly as it is, it seems we have to check for all the possible types here.

The problem has been noticed with AddressSanitizer:
```
W	6	==4830==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x60e0001e16e0 in thread T0: task="alio2-cr1-hv-gw01.cern.ch:/opt/git/ControlWorkflows/tasks/jit-cd2f2834da7ac943a17d6874dec53f825d9865a5-qc-pp-FDD-Quality@00552903229b92bff98c32ea79a875b8379a5e31#2oPDp5Pdrqa"
W	6	object passed to delete has wrong type: task="alio2-cr1-hv-gw01.cern.ch:/opt/git/ControlWorkflows/tasks/jit-cd2f2834da7ac943a17d6874dec53f825d9865a5-qc-pp-FDD-Quality@00552903229b92bff98c32ea79a875b8379a5e31#2oPDp5Pdrqa"
W	6	size of the allocated type: 152 bytes; task="alio2-cr1-hv-gw01.cern.ch:/opt/git/ControlWorkflows/tasks/jit-cd2f2834da7ac943a17d6874dec53f825d9865a5-qc-pp-FDD-Quality@00552903229b92bff98c32ea79a875b8379a5e31#2oPDp5Pdrqa"
W	6	size of the deallocated type: 96 bytes. task="alio2-cr1-hv-gw01.cern.ch:/opt/git/ControlWorkflows/tasks/jit-cd2f2834da7ac943a17d6874dec53f825d9865a5-qc-pp-FDD-Quality@00552903229b92bff98c32ea79a875b8379a5e31#2oPDp5Pdrqa"
```

knopers8 committed:
CCDB: cast to the concrete pointer type before deleting a uv_handle_t*

libuv handlers use c-style polymorphism, so deleting uv_handle_t* pointing to a derived class will in fact delete wrong number of bytes.
As ugly as it is, it seems we have to check for all the possible types here.

ktf commented on CCDB/src/CCDBDownloader.cxx:
```diff
@@ -136,7 +136,54 @@ void closeHandles(uv_handle_t* handle, void* arg)
 void onUVClose(uv_handle_t* handle)
 {
   if (handle != nullptr) {
-    delete handle;
```
Shouldn't you simply use malloc / free?

knopers8 replied:
Hm, indeed this could also work. I didn't remember that `free` should know how much bytes to delete. Let me try...

knopers8 replied:
seems also OK

knopers8 committed:
CCDB: use malloc/free for uv_handle_t to release the correct size on heap

While the previous approach was probably also OK, it's indeed more concise with C...
Asan output for testCcdbApiDownloader is now perfectly clean.

knopers8 commented:
@ktf , @costing , @sawenzel could one of you have a look?

costing commented:
Adding @TrifleMichael for information and to comment on it.

sawenzel commented:
@knopers8 : Fine for me. I assume the AddressSanitizer error is gone with this PR?

knopers8 commented:
> @knopers8 : Fine for me. I assume the AddressSanitizer error is gone with this PR?

Yes
